src/main/java/com/example/Restaurant_Management_System_REST_API/service/ReportService.java

```diff
@@ -0,0 +1,392 @@
+import org.apache.poi.ss.usermodel.Workbook;
+import org.apache.poi.xssf.usermodel.XSSFWorkbook;
+import org.modelmapper.ModelMapper;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.stereotype.Service;
+import org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBody;
+
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Service
+@AllArgsConstructor
+public class ReportService {
+
+    private final InventoryItemRepository inventoryItemRepository;
+    private final CustomerRepository customerRepository;
+    private final MenuRecordRepository menuRecordRepository;
+    private final ReservationRepository reservationRepository;
+    private final RestaurantOrderRepository restaurantOrderRepository;
+    private final SupplierRepository supplierRepository;
+    private final TableRepository tableRepository;
+    private final ModelMapper modelMapper;
+    private final RestaurantOrderMenuRecordRepository restaurantOrderMenuRecordRepository;
+    private final CellValueSetter<String> stringSetter = new StringValueSetter();
+    private final CellValueSetter<Long> longSetter = new LongValueSetter();
+    private final CellValueSetter<Double> doubleSetter = new DoubleValueSetter();
+
+    //section for InventoryItem reports
+    public StreamingResponseBody getInventoryItemByAmountGreaterThan(double amount) {
+
+        List<InventoryItem> items = inventoryItemRepository.findByAmountGreaterThan(amount);
+
+        // The StreamingResponseBody is used to stream the response back to the client. This is particularly useful for large files which can’t be held in memory.
+
+        //Below I create an excel file using Apache POI library and then I write it to outputStream for the client to download it
+        StreamingResponseBody stream = outputStream -> {
+            Workbook workbook = new XSSFWorkbook(); // a top level object to create sheets and other operations
+            Sheet sheet = workbook.createSheet("InventoryItems");
+
+            // Create header row
+            Row headerRow = sheet.createRow(0);
+            stringSetter.setCellValue(headerRow, 0, "ID");
+            stringSetter.setCellValue(headerRow, 1, "Name");
+            stringSetter.setCellValue(headerRow, 2, "Description");
+            stringSetter.setCellValue(headerRow, 3, "Price");
+            stringSetter.setCellValue(headerRow, 4, "Amount");
+            stringSetter.setCellValue(headerRow, 5, "Supplier Name");
+            stringSetter.setCellValue(headerRow, 6, "Supplier Street");
+            stringSetter.setCellValue(headerRow, 7, "Supplier House Number");
+            stringSetter.setCellValue(headerRow, 8, "Supplier City");
+            stringSetter.setCellValue(headerRow, 9, "Supplier Postal Code");
+            stringSetter.setCellValue(headerRow, 10, "Supplier Telephone Number");
+
```

Comment:
This part of code can be easily improved by using a simple loop. Could you take a look also at other places where this pattern of repeating lines continues?

Reply:
I see the point but because there are so many methods and almost each has different names of rows etc I dont see it is so easy...

Reply:
so this is how it might look like:

public StreamingResponseBody getInventoryItemByAmountGreaterThan(double amount) {
        List<InventoryItem> items = inventoryItemRepository.findByAmountGreaterThan(amount);
        return generateInventoryItemsReport(items);
    }

    public StreamingResponseBody getCustomerByRole(String roleName) {
        List<Customer> customerByRole = customerRepository.findByAuthorities_Authority_name(roleName);
        return generateCustomerByRoleReport(customerByRole);
    }

    public StreamingResponseBody getRestaurantOrderByOrderTimeRange(LocalDate startDate, LocalDate endDate) {
        List<RestaurantOrder> orders = restaurantOrderRepository.findByOrderTimeRange(startDate.atStartOfDay(), endDate.plusDays(1).atStartOfDay());
        return generateRestaurantOrderReport(orders, "restaurantOrdersByTimeRange");
    }

    public StreamingResponseBody getRestaurantOrderByOrderStatus(OrderStatus orderStatus) {
        List<RestaurantOrder> orders = restaurantOrderRepository.findByOrderStatus(orderStatus);
        return generateRestaurantOrderWithMenuRecordsReport(orders);
    }

    public StreamingResponseBody getRestaurantOrderMenuRecordInTimePeriod(LocalDate dateFrom, LocalDate dateTo) {
        List<RestaurantOrderMenuRecord> records = restaurantOrderMenuRecordRepository.findRestaurantOrderMenuRecordByTimePeriod(dateFrom.atStartOfDay(), dateTo.plusDays(1).atStartOfDay());
        return generatePopularDishesReport(records, dateFrom, dateTo);
    }

    private StreamingResponseBody generateInventoryItemsReport(List<InventoryItem> items) {
        return outputStream -> {
            try (Workbook workbook = new XSSFWorkbook()) {
                Sheet sheet = workbook.createSheet("InventoryItems");
                createHeaderRow(sheet, ID, NAME, DESCRIPTION, PRICE, AMOUNT, SUPPLIER_NAME, SUPPLIER_STREET, SUPPLIER_HOUSE_NUMBER, SUPPLIER_CITY, SUPPLIER_POSTAL_CODE, SUPPLIER_TELEPHONE_NUMBER);

                for (int i = 0; i < items.size(); i++) {
                    InventoryItem item = items.get(i);
                    Supplier supplier = item.getSupplier();
                    ContactDetails supplierContactDetails = supplier.getContactDetails();
                    Row row = sheet.createRow(i + 1);
                    populateRowWithInventoryItem(row, item, supplierContactDetails);
                }

                workbook.write(outputStream);
            }
        };
    }

    private StreamingResponseBody generateCustomerByRoleReport(List<Customer> customers) {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern(DATE_FORMAT);
        return outputStream -> {
            try (Workbook workbook = new XSSFWorkbook()) {
                Sheet sheet = workbook.createSheet("customersByRoles");
                createHeaderRow(sheet, CUSTOMER_ID, CREATION_TIME, RESERVATION_ID, CUSTOMER_NAME, ACCOUNT_ENABLED, ROLES);

                for (int i = 0; i < customers.size(); i++) {
                    Customer customer = customers.get(i);
                    Row row = sheet.createRow(i + 1);
                    populateRowWithCustomer(row, customer, formatter);
                }

                workbook.write(outputStream);
            }
        };
    }

    private StreamingResponseBody generateRestaurantOrderReport(List<RestaurantOrder> orders, String sheetName) {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern(DATE_FORMAT);
        return outputStream -> {
            try (Workbook workbook = new XSSFWorkbook()) {
                Sheet sheet = workbook.createSheet(sheetName);
                createHeaderRow(sheet, RESTAURANT_ORDER_ID, ORDER_TIME, ORDER_STATUS, TABLE_ID, CUSTOMER_TELEPHONE_NUMBER, ORDER_TOTAL_AMOUNT_TO_PAY);

                for (int i = 0; i < orders.size(); i++) {
                    RestaurantOrder order = orders.get(i);
                    Row row = sheet.createRow(i + 1);
                    populateRowWithRestaurantOrder(row, order, formatter);
                }

                workbook.write(outputStream);
            }
        };
    }

    private StreamingResponseBody generateRestaurantOrderWithMenuRecordsReport(List<RestaurantOrder> orders) {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern(DATE_FORMAT);
        return outputStream -> {
            try (Workbook workbook = new XSSFWorkbook()) {
                Sheet sheet = workbook.createSheet("restaurantOrdersByOrderStatus");
                createHeaderRow(sheet, RESTAURANT_ORDER_ID, ORDER_TIME, ORDER_STATUS, TABLE_ID, CUSTOMER_TELEPHONE_NUMBER, ORDER_TOTAL_AMOUNT_TO_PAY, MENU_RECORD_NAME, PORTIONS_AMOUNT);

                int rowIndex = 1;
                for (RestaurantOrder order : orders) {
                    List<RestaurantOrderMenuRecord> menuRecords = order.getRestaurantOrders();
                    for (RestaurantOrderMenuRecord record : menuRecords) {
                        Row row = sheet.createRow(rowIndex++);
                        populateRowWithRestaurantOrderMenuRecord(row, order, record, formatter);
                    }
                }

                workbook.write(outputStream);
            }
        };
    }

    private StreamingResponseBody generatePopularDishesReport(List<RestaurantOrderMenuRecord> records, LocalDate dateFrom, LocalDate dateTo) {
        return outputStream -> {
            try (Workbook workbook = new XSSFWorkbook()) {
                Sheet sheet = workbook.createSheet("popularDishes");
                createPeriodRow(sheet, dateFrom, dateTo);

                Map<String, Double> dishPortionsMap = new HashMap<>();
                for (RestaurantOrderMenuRecord record : records) {
                    String dishName = record.getMenuRecord().getName();
                    Double portions = record.getPortionsAmount();
                    dishPortionsMap.put(dishName, dishPortionsMap.getOrDefault(dishName, 0.0) + portions);
                }

                int rowIndex = 2;
                for (Entry<String, Double> entry : dishPortionsMap.entrySet()) {
                    Row row = sheet.createRow(rowIndex++);
                    stringSetter.setCellValue(row, 0, entry.getKey());
                    doubleSetter.setCellValue(row, 1, entry.getValue());
                }

                workbook.write(outputStream);
            }
        };
    }

    private void createHeaderRow(Sheet sheet, String... headers) {
        Row headerRow = sheet.createRow(0);
        for (int i = 0; i < headers.length; i++) {
            stringSetter.setCellValue(headerRow, i, headers[i]);
        }
    }

    private void createPeriodRow(Sheet sheet, LocalDate dateFrom, LocalDate dateTo) {
        Row periodRow = sheet.createRow(0);
        stringSetter.setCellValue(periodRow, 0, "Time period: " + dateFrom + " - " + dateTo);
        createHeaderRow(sheet, MENU_RECORD_NAME, PORTIONS_AMOUNT);
    }

    private void populateRowWithInventoryItem(Row row, InventoryItem item, ContactDetails supplierContactDetails) {
        longSetter.setCellValue(row, 0, item.getId());
        stringSetter.setCellValue(row, 1, item.getName());
        stringSetter.setCellValue(row, 2, item.getDescription());
        doubleSetter.setCellValue(row, 3, item.getPrice());
        doubleSetter.setCellValue(row, 4, item.getAmount());
        stringSetter.setCellValue(row, 5, supplierContactDetails.getName());
        stringSetter.setCellValue(row, 6, supplierContactDetails.getStreet());
        stringSetter.setCellValue(row, 7, supplierContactDetails.getHouseNumber());
        stringSetter.setCellValue(row, 8, supplierContactDetails.getCity());
        stringSetter.setCellValue(row, 9, supplierContactDetails.getPostalCode());
        stringSetter.setCellValue(row, 10, supplierContactDetails.getTelephoneNumber());
    }

    private void populateRowWithCustomer(Row row, Customer customer, DateTimeFormatter formatter) {
        longSetter.setCellValue(row, 0, customer.getId());
        stringSetter.setCellValue(row, 1, customer.getCreationTime().format(formatter));
        if (customer.getReservation() != null) {
            longSetter.setCellValue(row, 2, customer.getReservation().getId());
        }
        stringSetter.setCellValue(row, 3, customer.getContactDetails().getName());
        stringSetter.setCellValue(row, 4, customer.getEnabled().toString());
        Collection<? extends GrantedAuthority> authorities = customer.getAuthorities();
        String roles = authorities.stream()
                .map(GrantedAuthority::getAuthority)
                .collect(Collectors.joining(", "));
        stringSetter.setCellValue(row, 5, roles);
    }

    private void populateRowWithRestaurantOrder(Row row, RestaurantOrder order, DateTimeFormatter formatter) {
        longSetter.setCellValue(row, 0, order.getId());
        stringSetter.setCellValue(row, 1, order.getOrderTime().format(formatter));
        stringSetter.setCellValue(row, 2, order.getOrderStatus().toString());
        longSetter.setCellValue(row, 3, order.getTable().getId());
        stringSetter.setCellValue(row, 4, order.getTelephoneNumber());
        doubleSetter.setCellValue(row, 5, order.getTotalAmountToPay());
    }

    private void populateRowWithRestaurantOrderMenuRecord(Row row, RestaurantOrder order, RestaurantOrderMenuRecord record, DateTimeFormatter formatter) {
        longSetter.setCellValue(row, 0, order.getId());
        stringSetter.setCellValue(row, 1, order.getOrderTime().format(formatter));
        stringSetter.setCellValue(row, 2, order.getOrderStatus().toString());
        longSetter.setCellValue(row, 3, order.getTable().getId());
        stringSetter.setCellValue(row, 4, order.getTelephoneNumber());
        doubleSetter.setCellValue(row, 5, order.getTotalAmountToPay());
        stringSetter.setCellValue(row, 6, record.getMenuRecord().getName());
        doubleSetter.setCellValue(row, 7, record.getPortionsAmount());
    }
}

Reply:
it works perfectly - I checked - I explained the problem to AI and it proposed such code. I believe this is what you expect but is that so easy to read for me ? hmmm. I understand it but because this is my first time when I learned how to generate an excel files as reports I think I would rather to stay with the old version of code - the one which is currently in this pr. Why? Its more readable for me for now.

Reply:
I'm not sure the AI understood what it was given as an input :) To me it looks like it didn't improve anything. I just meant to rework the code and obey the DRY rule by using loop, i.e.:

```java
String[] headers = {
    "ID", "Name", "Description", "Price", "Amount",
    "Supplier Name", "Supplier Street", "Supplier House Number",
    "Supplier City", "Supplier Postal Code", "Supplier Telephone Number"
};

Row headerRow = sheet.createRow(0);

for (int i = 0; i < headers.length; i++) {
    stringSetter.setCellValue(headerRow, i, headers[i]);
}
```

which looks way better IMO. That's improving readability and cleanliness of the code base.

Concepts like that can be applied in this PR in a few places which I'd like to challenge you with finding them yourself as a way to further expand the skillset :)

Reply:
yes I know what you mean, AI also did it and it went in these changes even further on.
But for now refactoring into arrays is I think enough and I make those right now!